### PR TITLE
Query Fixer: reorganize the previous PR and init the project

### PR DESCRIPTION
**Pull request by mingen-pan**

### CODE REVIEW

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessor.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  private final BigQueryParserFactory parserFactory;
+
+  /**
+   * tokenize a query and return all its tokens.
+   * @param sql the query to be tokenized
+   * @return a list of tokens of the query
+   */
+  public List<IToken> getAllTokens(String sql) {
+    Objects.requireNonNull(sql, "input query should not be null");
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessor.java:30]: preconditions.

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessor.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    List<IToken> tokens = new ArrayList<>();
+
+    // SqlBabelParserImpl has a token manager to tokenize the input query.
+    SqlBabelParserImpl parserImpl = parserFactory.getBabelParserImpl(sql);
+
+    Token token;
+    final int EndKind = 0;
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessor.java:38]: If it is going to used in other places, you can make it static const of the class.

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/IToken.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package com.google.cloud.bigquery.utils.queryfixer.entity;
+
+/**
+ * A interface for queries' tokens.
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/IToken.java:4]: nit: "**An** interface"

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/TokenImpl.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @Override public String toString() {
+    return  String.format("%s [%d:%d]", token.image, token.beginLine, token.beginColumn);
+  }
+
+  //  public static TokenImpl newToken(int ofKind) {
+//    return new TokenImpl(Token.newToken(ofKind));
+//  }
+}
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/TokenImpl.java:52]: remove the commented line?

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/exception/ParserCreationException.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.google.cloud.bigquery.utils.queryfixer.exception;
+
+
+public class ParserCreationException extends RuntimeException {
+  public ParserCreationException(String message) {
+    super(message);
+  }
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/exception/ParserCreationException.java:7]: Add  a comment for the exception?

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/IToken.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * */
+public interface IToken {
+
+  int getKind();
+
+  String getImage();
+
+  int getBeginLine();
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/IToken.java:12]: nit: getBeginRow(); getBeginColumn()

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/TokenImpl.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  private final Token token;
+
+  public TokenImpl (Token token) {
+    this.token = token;
+  }
+
+  @Override public int getKind() {
+    return token.kind;
+  }
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/entity/TokenImpl.java:18]: nit: not sure this has been auto formatted yet... i think usually this should be
@Override
public int getKind() {
...
}

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import static org.junit.Assert.assertEquals;
+
+public class QueryTokenProcessorTest {
+
+  private QueryTokenProcessor createService() {
+    return new QueryTokenProcessor(new BigQueryParserFactory());
+  }
+
+  @Test
+  public void convertSqlToTokens() {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:18]: nit: convertQueryToTokens()

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class QueryTokenProcessorTest {
+
+  private QueryTokenProcessor createService() {
+    return new QueryTokenProcessor(new BigQueryParserFactory());
+  }
+
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:16]: This should be shared by all test cases in setup()?

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  @Test
+  public void convertSqlToTokens() {
+    QueryTokenProcessor tokenService = createService();
+    String sql = "Select col from `d1.t1`\n" +
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:20]: The auto formatting seems a bit weird. The line is too short...

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        "where t1.col>'val'";
+
+    List<IToken> tokens = tokenService.getAllTokens(sql);
+    assertEquals(10, tokens.size());
+    for (IToken token : tokens) {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:25]: Usually we shouldn't print out these in a test. 
If we want to test out the contents, we should use List<IToken> expectedTokens  = ... and assert if they are equal

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package com.google.cloud.bigquery.utils.queryfixer;
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:1]: nit: not sure why there are so many empty lines in between...

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+  @Test
+  public void getParser() throws SqlParseException {
+    BigQueryParserFactory factory = new BigQueryParserFactory();
+    SqlParser parser = factory.getParser("select a from b");
+    SqlNode node = parser.parseQuery();
+    assertEquals(SqlKind.SELECT, node.getKind());
+  }
+
+  @Test
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java:26]: The best practice for writing tests:

Tests independent behaviors of a method separately:
Example: 
@Test
public void displayTransactionResults_showsItemName() {
  transactionProcessor.displayTransactionResults(
      new User(), new Transaction("Some Item"));
  assertThat(ui.getText()).contains("You bought a Some Item");
}

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+      factory.getBabelParserImpl(null);
+      fail();
+    } catch (NullPointerException e) {
+      assertEquals("the input query should not be null", e.getMessage());
+    }
+
+  }
+
+  @Test
+  public void getParserImplFromIncorrectFactory() {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java:40]: public void getParserImpl_returnsIncorrectParser() {}

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class ParseFactoryTest {
+
+  @Test
+  public void getParser() throws SqlParseException {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java:19]: Can we add a case for a slightly more complicated query?

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    return SqlParser.create(source, parserConfig);
+  }
+
+  private SqlParser.Config buildConfig(SqlParserImplFactory factory) {
+    final SqlParser.ConfigBuilder configBuilder =
+        SqlParser.configBuilder()
+            .setParserFactory(factory)
+            .setQuoting(quoting)
+            .setUnquotedCasing(unquotedCasing)
+            .setQuotedCasing(quotedCasing)
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:87]: If most of the settings are default settings, we should consider building it in statically and only set factory here.

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 9 above)
Comment (kikkyo) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:16]: also createProcessor() :)

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public class BigQueryParserFactory {
+
+  private static final Quoting quoting = Quoting.BACK_TICK;
+  private static final Casing quotedCasing = Casing.UNCHANGED;
+  private static final SqlConformance conformance = SqlConformanceEnum.DEFAULT;
+  private final SqlParser.Config parserConfig;
+  private static final Casing unquotedCasing = Casing.UNCHANGED;
+
+  /** Default initialization with the Babel Parser factory. */
```
[the remark this answers is not in the document]
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:29]: I think it is good to mention the default type pf the parser is Babel.

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java (hunk 15 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java]: A new case is added.

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 12 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:1]: I think it is caused by auto format.

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java (hunk 14 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParseFactoryTest.java]: Done

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 10 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java]: I break this on purpose. I want to simulate the BigQuery console, where a user can type a query in multiple lines.

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 11 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:24]: Done

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.calcite.util.SourceStringReader;
+import org.apache.commons.lang3.reflect.FieldUtils;
+
+import com.google.cloud.bigquery.utils.queryfixer.exception.ParserCreationException;
+
+import java.io.Reader;
+import java.util.Objects;
+
+/**
+ * A factory to generate parsers. The fault generated parser is Babel Parser with BigQuery dialect.
```
[the remark this answers is not in the document]
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:19]: Done

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * Get the implementation of a parser parsing a query. The implementation provides functions to
+   * tokenize the query.
+   *
+   * @param query the query fed to the parser implementation.
+   * @return SqlBabelParserImpl the parser implementation
+   */
+  public SqlBabelParserImpl getBabelParserImpl(String query) {
```
[the remark this answers is not in the document]
Comment (kikkyo) [reply tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:60]: ok.thx!

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (kikkyo) [reply tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:19]: looks like we still use "fault" :)

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 9 above)
Comment (mingen-pan) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:16]: Miss this comment. I will revise this in next commit

File: tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java [hunk 20]
```diff
@@ -58,18 +58,19 @@ public SqlParser getParser(String sql) {
    * @return SqlBabelParserImpl the parser implementation
    */
   public SqlBabelParserImpl getBabelParserImpl(String query) {
-    Objects.requireNonNull(query, "the input query should not be null");
+    Preconditions.checkNotNull(query, "The input query should not be null.");
 
     Object parserImpl;
     try {
-      parserImpl = FieldUtils.readField(getParser(query), "parser", true);
+      parserImpl = FieldUtils.readField(getParser(query), /* fieldName= */"parser", /* forceAccess= */true);
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/BigQueryParserFactory.java:65]: nit: There should be space in between /* forceAccess= */<extra space*> true

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java [hunk 21]
```diff
@@ -11,18 +11,29 @@
...
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
-public class ParseFactoryTest {
+public class ParserFactoryTest {
 
   @Test
-  public void getParser() throws SqlParseException {
+  public void getParser1() throws SqlParseException {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java:20]: getParser_selectQuery()

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java [hunk 22]
```diff
@@ -11,18 +11,29 @@
...
     BigQueryParserFactory factory = new BigQueryParserFactory();
     SqlParser parser = factory.getParser("select a from b");
     SqlNode node = parser.parseQuery();
     assertEquals(SqlKind.SELECT, node.getKind());
   }
 
+  @Test
+  public void getParser2() throws SqlParseException {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java:28]: getParser_selectWithSubQuery()

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java [hunk 23]
```diff
@@ -11,18 +11,29 @@
...
+    BigQueryParserFactory factory = new BigQueryParserFactory();
+    SqlParser parser =
+        factory.getParser(
+            "select * from (select a, b, c from x cross join y where d > 10) inner join z");
+    SqlNode node = parser.parseQuery();
+    assertEquals(SqlKind.SELECT, node.getKind());
+  }
+
   @Test
   public void queryNotNull() {
```
Comment (kikkyo) [inline tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/ParserFactoryTest.java:38]: getParser_nullQuery

File: tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java (hunk 9 above)
Comment (kikkyo) [reply tools/automatic_query_fixer/src/test/java/com/google/cloud/bigquery/utils/queryfixer/QueryTokenProcessorTest.java:16]: I think you can make it a member in testSetup() so you don't have to call it for every test case

**Pull request merged**